### PR TITLE
[2.9] irmin: use CPS for conversions between trees and proofs

### LINKED ISSUE #1600: Merkle Proof computation not CPS
Merkle Proof computation is not written in CPS, is that a problem?

### PR DESCRIPTION
**Pull request by samoht**
On top of #1623 

Fixes #1600

### CODE REVIEW

File: src/irmin/tree.ml [hunk 1]
```diff
@@ -1841,69 +1848,93 @@ module Make (P : Private.S) = struct
...
+        | [] -> k (`Node (Node.of_map acc))
+        | (s, p) :: rest -> to_tree p (fun n -> aux (StepMap.add s n acc) rest)
       in
-      `Node (Node.of_map bindings)
+      aux StepMap.empty n
 
     (** [tree_of_inode] is solely called on the root of an inode tree *)
-    and tree_of_inode len proofs : tree =
-      let elts =
+    and tree_of_inode : type a. int -> (int * t) list -> (tree -> a) -> a =
```
Comment (Ngoguey42) [inline src/irmin/tree.ml:1871]: This level of recursion is not in CPS, but I guess this is alright

**Review by Ngoguey42 — APPROVED** [review]

File: src/irmin/tree.ml (hunk 1 above)
Comment (samoht) [reply src/irmin/tree.ml:1871]: Fixed in e9d9b0b

Comment (Ngoguey42) [reply src/irmin/tree.ml:1871]: That's better. However, the `tree_of_inode` frame still remains during de computation of the subtrees. The the last 2 lines of `tree_of_inode` should be in the CPS callback:
```
      List.iter (fun (s, elt) -> Node.add_to_findv_cache n s elt) elts;
      k (`Node n)
```

Comment (samoht) [reply src/irmin/tree.ml:1871]: thanks, should be fixed in e9d9b0b

Comment (Ngoguey42) [reply src/irmin/tree.ml:1871]: LGTM

**Pull request merged**